src/txrequest.h

```diff
@@ -0,0 +1,216 @@
+ * - Which peer announced it (through their NodeId)
+ * - The txid or wtxid of the transaction (collectively called "txhash" in what follows)
+ * - Whether it was a tx or wtx announcement (see BIP339).
+ * - What the earliest permitted time is that that transaction can be requested from that peer (called "reqtime").
+ * - Whether it's from a "preferred" peer or not. Which announcements get this flag is determined by the caller, but
+ *   this is designed for outbound peers, or other peers that we have a higher level of trust in). Even when the
+ *   peers' preferredness changes, the preferred flag of existing announcements from that peer won't change.
+ * - Whether the peer was the "first" to announce this txhash within its class (see '"First" marker rules').
+ * - Whether or not the transaction was requested already, and if so, when it times out (called "expiry").
+ * - Whether or not the transaction request failed already (timed out, or NOTFOUND was received).
+ *
+ * Transaction requests are then assigned to peers, following these rules:
+ *
+ * - No transaction is requested as long as another request for the same txhash is outstanding (it needs to fail
+ *   first by passing expiry, or a NOTFOUND or invalid transaction has to be received for it).
+ *
+ *   Rationale: to avoid wasting bandwidth on multiple copies of the same transaction.
+ *
+ * - The same transaction is never requested twice from the same peer, unless the transaction was forgotten in
+ *   between (see next point), and re-announced.
+ *
+ *   Rationale: giving a peer multiple chances to announce a transaction would allow them to bias requests in their
+ *              favor, worsening invblock attacks. The flip side is that as long as an attacker manages to prevent
+ *              us from receiving a transaction, failed announcements (including those from honest peers) will
+ *              linger longer, increasing memory usage somewhat. The impact of this is limited by imposing a cap on
+ *              the number of tracked announcements per peer.
+ *              Invblocking is the practice of announcing transactions but not answer requests for them, in order
+ *              to delay (or prevent) a target learning the transaction. See
+ *              https://allquantor.at/blockchainbib/pdf/miller2015topology.pdf for more information.
+ *
+ * - Announcements are only forgotten about when the peer that announced them went offline, when the transaction
+ *   was received successfully, or when no candidates for a transaction remain that haven't been tried already.
+ *
+ *   Rationale: we need to eventually forget announcements to keep memory bounded, but as long as viable
+ *              candidate peers remain, we prefer to avoid fetching from failed ones. As every request has a finite
+ *              timeout and we schedule new request as soon a previous one expired, there is always progress being
```

Comment:
"progress towards forgetting a transaction" is unambiguous I think? Every minute one peer is crossed off the list of candidates, so eventually you will run out. Note that this section isn't about successfully receiving the transaction, but bounding memory usage.

Reply:
Actually, thinking more about this, it isn't exactly true, because for non-preferred connections we should assume the attacker can disconnect and reconnect, giving them a new opportunity. I think this should be restated as: an attacker can force us to keep a transaction in memory (even in the queues of honest peers) for as long as they can prevent us from receiving the transaction. That's not necessarily bounded, but in cases where it isn't, we have bigger problems. Actual OOM is prevented using per-peer queue size limits.

Reply:
I've adjusted the comments, merging this paragraph into the previous one (they were already kind of circularly referring to each other).